Willow MSBuild does not look up MSBuild extension path under VS folder when using MSBuildExtensionPath32

Running from a Willow command-prompt, with the following variations of imports:
``` XML
  <VSToolsPath Condition="'$(VSToolsPath)' == ''">$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)</VSToolsPath>
[...]
   <Import Project="$(VSToolsPath)\VSSDK\Microsoft.VsSDK.targets"" />
```

``` XML
<Import Project="$(VSToolsPath)\VSSDK\Microsoft.VsSDK.targets" />
```

and

``` XML
<Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\VSSDK\Microsoft.VsSDK.targets"/>
```

Under a Willow command-prompt, I would expect all of these to get found under C:\Program Files (x86)\Microsoft Visual Studio 15.0\MSBuild\Microsoft\VisualStudio\v15.0\VSSDK.

Instead I get:
``` 
D:\roslyn-project-system\build\Targets\VSL.Imports.targets(368,3): error MSB4019: The imported project "C:\Program File
s (x86)\MSBuild\Microsoft\VisualStudio\v15.0\VSSDK\Microsoft.VsSDK.targets" was not found. Confirm that the path in the
 <Import> declaration is correct, and that the file exists on disk. [D:\roslyn-project-system\src\ProjectSystemDogfoodS
etup\ProjectSystemDogfoodSetup.csproj] [D:\roslyn-project-system\build\build.proj]
```

Oh, I looked at the log - turns out this is NuGet.exe running MSBuild: 

```
                   Using "Exec" task from assembly "Microsoft.Build.Tasks.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a".
                   Task "Exec" (TaskId:3)
                     Task Parameter:Command="C:\roslyn-project-system\build\..\build\bin\NuGet.exe" restore -verbosity quiet "C:\roslyn-project-system\build\..\src\ProjectSystem.sln" (TaskId:3)
                     "C:\roslyn-project-system\build\..\build\bin\NuGet.exe" restore -verbosity quiet "C:\roslyn-project-system\build\..\src\ProjectSystem.sln" (TaskId:3)
18:34:52.092     1>C:\roslyn-project-system\build\Targets\VSL.Imports.targets(365,3): error MSB4019: The imported project "C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v15.0\VSSDK\Microsoft.VsSDK.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [C:\roslyn-project-system\src\Microsoft.VisualStudio.ProjectSystem.CSharp.VS\Microsoft.VisualStudio.ProjectSystem.CSharp.VS.csproj] [C:\roslyn-project-system\build\build.proj]
18:34:52.190     1>C:\roslyn-project-system\build\Targets\VSL.Imports.targets(365,3): error MSB4019: The imported project "C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v15.0\VSSDK\Microsoft.VsSDK.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [C:\roslyn-project-system\src\Microsoft.VisualStudio.ProjectSystem.VisualBasic.VS\Microsoft.VisualStudio.ProjectSystem.VisualBasic.VS.csproj] [C:\roslyn-project-system\build\build.proj]
18:34:52.304     1>C:\roslyn-project-system\build\Targets\VSL.Imports.targets(365,3): error MSB4019: The imported project "C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v15.0\VSSDK\Microsoft.VsSDK.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [C:\roslyn-project-system\src\Microsoft.VisualStudio.ProjectSystem.CSharp.VS\Microsoft.VisualStudio.ProjectSystem.CSharp.VS.csproj] [C:\roslyn-project-system\build\build.proj]
18:34:52.322     1>C:\roslyn-project-system\build\Targets\VSL.Imports.targets(365,3): error MSB4019: The imported project "C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v15.0\VSSDK\Microsoft.VsSDK.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [C:\roslyn-project-system\src\Microsoft.VisualStudio.ProjectSystem.CSharp.VS\Microsoft.VisualStudio.ProjectSystem.CSharp.VS.csproj] [C:\roslyn-project-system\build\build.proj]
18:34:52.438     1>C:\roslyn-project-system\build\Targets\VSL.Imports.targets(365,3): error MSB4019: The imported project "C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v15.0\VSSDK\Microsoft.VsSDK.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [C:\roslyn-project-system\src\Microsoft.VisualStudio.ProjectSystem.VisualBasic.VS\Microsoft.VisualStudio.ProjectSystem.VisualBasic.VS.csproj] [C:\roslyn-project-system\build\build.proj]
18:34:52.458     1>C:\roslyn-project-system\build\Targets\VSL.Imports.targets(365,3): error MSB4019: The imported project "C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v15.0\VSSDK\Microsoft.VsSDK.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [C:\roslyn-project-system\src\Microsoft.VisualStudio.ProjectSystem.VisualBasic.VS\Microsoft.VisualStudio.ProjectSystem.VisualBasic.VS.csproj] [C:\roslyn-project-system\build\build.proj]
18:34:52.472     1>C:\roslyn-project-system\build\Targets\VSL.Imports.targets(365,3): error MSB4019: The imported project "C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v15.0\VSSDK\Microsoft.VsSDK.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [C:\roslyn-project-system\src\ProjectSystemSetup\ProjectSystemSetup.csproj] [C:\roslyn-project-system\build\build.proj]
18:34:52.480     1>C:\roslyn-project-system\build\Targets\VSL.Imports.targets(365,3): error MSB4019: The imported project "C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v15.0\VSSDK\Microsoft.VsSDK.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [C:\roslyn-project-system\src\VisualStudioEditorsSetup\VisualStudioEditorsSetup.csproj] [C:\roslyn-project-system\build\build.proj]
18:34:52.631     1>C:\roslyn-project-system\build\Targets\VSL.Imports.targets(365,3): error MSB4019: The imported project "C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v15.0\VSSDK\Microsoft.VsSDK.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [C:\roslyn-project-system\src\ProjectSystemDogfoodSetup\ProjectSystemDogfoodSetup.csproj] [C:\roslyn-project-system\build\build.proj]
                      (TaskId:3)
18:34:52.674     1>C:\roslyn-project-system\build\build.proj(30,5): error MSB3073: The command ""C:\roslyn-project-system\build\..\build\bin\NuGet.exe" restore -verbosity quiet "C:\roslyn-project-system\build\..\src\ProjectSystem.sln"" exited with code 1.
                   Done executing task "Exec" -- FAILED. (TaskId:3)
18:34:52.674     1>Done building target "RestorePackages" in project "build.proj" -- FAILED.: (TargetId:2)
18:34:52.675     1>Done Building Project "C:\roslyn-project-system\build\build.proj" (Build target(s)) -- FAILED.
```

--- a/src/XMakeTasks/CodeTaskFactory.cs
+++ b/src/XMakeTasks/CodeTaskFactory.cs
@@ -31,6 +31,37 @@ namespace Microsoft.Build.Tasks
     public class CodeTaskFactory : ITaskFactory
     {
         /// <summary>
+        /// This dictionary keeps track of custom references to compiled assemblies.  The in-memory assembly is loaded from a byte
+        /// stream and as such its dependencies cannot be found unless they are in the MSBuild.exe directory or the GAC.  They
+        /// cannot be found even if they are already loaded in the AppDomain.  This dictionary knows the FullName of the assembly
+        /// and a reference to the assembly itself.  In the <see cref="CurrentDomainOnAssemblyResolve"/> handler, the dictionary 
+        /// is used to return the loaded assemblies as a way to allow custom references that are not in the normal assembly Load
+        /// context.
+        /// </summary>
+        private static readonly IDictionary<string, Assembly> s_knownReferenceAssemblies = new ConcurrentDictionary<string, Assembly>(StringComparer.OrdinalIgnoreCase);
+
+        static CodeTaskFactory()
+        {
+            // The handler is not detached because it only returns assemblies for custom references that cannot be found in the normal Load context
+            AppDomain.CurrentDomain.AssemblyResolve += CurrentDomainOnAssemblyResolve;
+        }
+
+        /// <summary>
+        /// Handles the <see cref="AppDomain.AssemblyResolve"/> event to return assemblies loaded from custom references.
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="args"></param>
+        /// <returns></returns>
+        private static Assembly CurrentDomainOnAssemblyResolve(object sender, ResolveEventArgs args)
+        {
+            Assembly assembly = null;
+            // Return the assembly loaded from the custom reference if the FullName matches what is being looked for
+            s_knownReferenceAssemblies.TryGetValue(args.Name, out assembly);
+
+            return assembly;
+        }
+
+        /// <summary>
         /// Default assemblies names to reference during inline code compilation - from the .NET Framework
         /// </summary>
         private static readonly string[] s_defaultReferencedFrameworkAssemblyNames = { @"System.Core" };
@@ -657,6 +688,7 @@ namespace Microsoft.Build.Tasks
                                 if (candidateAssembly != null)
                                 {
                                     candidateAssemblyLocation = candidateAssembly.Location;
+                                    s_knownReferenceAssemblies.Add(candidateAssembly.FullName, candidateAssembly);
                                 }
                             }
                             catch (BadImageFormatException e)

--- a/src/XMakeTasks/CodeTaskFactory.cs
+++ b/src/XMakeTasks/CodeTaskFactory.cs
@@ -688,7 +688,7 @@ namespace Microsoft.Build.Tasks
                                 if (candidateAssembly != null)
                                 {
                                     candidateAssemblyLocation = candidateAssembly.Location;
-                                    s_knownReferenceAssemblies.Add(candidateAssembly.FullName, candidateAssembly);
+                                    s_knownReferenceAssemblies[candidateAssembly.FullName] = candidateAssembly;
                                 }
                             }
                             catch (BadImageFormatException e)

